### DISCUSSION TITLE
Stochastic frequency correction

### DISCUSSION THREAD

**Discussion by SunjunKim** [ExpressLRS/ExpressLRS#1704] [Ideas]
I came up with an idea for enabling frequency correction with the following advantages:
- At the expense of 1 LQ at most during the compensation in progress. No LQ drop is expected during normal operation.
- Measured from RX solely
- Without switching to the explicit header mode
- Automatic, without users' intervention
- Even possible with FLRC
- Simple and lightweight logic

Too good to be true? I think the following logic makes sense, but I want to hear your opinions as well.

Algorithm:
1. Have two "measurement" packets out of every 101 packets (the nearest prime number to 100). The measurement packets have -100 and +100 FREQ_STEP correction (=about 20 kHz shift)
2. Count the successful packet reception from those measurement packets. Let's say them CNT_LOW and CNT_HIGH. 
3. Wait for about 20 attempts (=2000 packets)
4. Compare CNT_LOW and CNT_HIGH. If LQ is good enough (>90) and they differ statistically significantly (for example, lost 15+ packets at one side), we assume TX/RX are at the marginal frequency error condition. If so, apply 50 `FreqCorrection` to compensate.
5. Repeat

The specific thresholds and compensation values should be tested and corrected, but I guess it's a good algorithm overall. It'll work slowly, but the frequency drift also doesn't change rapidly -- so I think it will work reasonably.

I'm planning to make PoC code, targeting v3.1.

**Comment by SunjunKim** [comment]
One possible drawback of this approach (and all other run-time freq compensation methods) will introduce some exciting situations like:

- I fly one pack normally
- On the next pack it won't bind!

For this reason, I also suggest having some warning notification to a user once the compensation above is activated.

**Reply by bike797** [reply]
Write down a few topics.
- RX frequency is a fixed value determined at the time of manufacture.
  Why is frequency correction always necessary (in flight)?
- Will frequency correction with temperature always help?
- What if it only works when frequency correction is required?
    Apart from the problem of not binding.
- Why not apply it only when necessary on the ground and not in flight?

**Reply by SunjunKim** [reply]
> * RX frequency is a fixed value determined at the time of manufacture.
>   Why is frequency correction always necessary (in flight)?

TX frequency shifts due to heat (e.g., https://youtu.be/hFRPYe1oMWY), and it could be substantial.

> * Will frequency correction with temperature always help?

Yes, but there's no systematic relationship between temperature and frequency shift, which prevents a simple compensation method. 

> * What if it only works when frequency correction is required?
>   Apart from the problem of not binding.

It's upon a principle that "once it binds, it should be OK". Not being able to bind from the first place is out of consideration of this solution.

> * Why not apply it only when necessary on the ground and not in flight?

TX frequency may shift over time, mostly a few tens of minutes.

**Comment by CapnBry** [comment]
Brilliant idea! Great to hear a new concept to consider instead of forcing tuning down a user's throat. I implemented this and spent the day playing with it. It definitely works, but obviously only if the receiver is so out of whack that it is about to fall off the edge of the tolerated range anyway. Some observations:

## Offset
I thought there would need to be a decent amount of shift for it to work, I so I started with `FreqCorrectionMax / 4` but after working with it, even `FreqCorrectionMax / 10` works fine, or 20kHz. This would only shift people whose `DEBUG_FREQ_CORRECTION` value is 114 or higher, which is pretty extreme. Some modes of course have tighter restrictions on the max error though, even different LoRa modes seem to have slightly different cutoffs.

## Frequency
The occurrence of the shifted packets has to be pretty common. I tried it first with one positive and one negative shift per Nonce cycle (once every 128 packets) and it was so very very slow even on 500Hz. It loops roughly twice a second, but you need a bunch of periods to get enough to form an opinion. I was waiting 20 shifted packets (10 high 10 low) and that's only moving it ~200Hz every 5 seconds. At 50Hz it would only be shifting every 50 seconds! I also tried shifting the frequency quite a bit (1kH - 10kHz) per cycle, but I think that is dangerous and would prefer a more gradual transition.

To accomplish this, I shift 2 out of every 8 packets, on slot 1 and slot 5 (0-based). These were chosen because telemetry can never occur on an odd slot so there's no need to track telemetry on the RX when looking for missing packets.

## Method
When a packet is received that passes CRC on the RX, the Nonce is stored for processing by the Synthetic Frequency Error Indicator (SynFei) process. The SynFei counts the number of packets received in slot 1 and slot 5. When the total count reaches some sum, the counts are compared and if they are different enough, the FreqCorrection is ticked one tick. With an 8 packet period, 20 total shifted packets seemed to work ok and the FreqCorrection was shifted if as few as 4 were missed by the opposite polarity. I used the total of the two counts as the trigger for when to do an adjustment as an attempt to any bias toward one polarity or the other, but I am not sure that makes any difference. It does make the reaction slower when the error is big, but speed up as it gets closer and more of the "missed" polarity packets start being received. This is good because once it gets kinda close, it naturally slows down so the overall pace of the adjustment stays more consistent.
```c
#define SYNFEI_PERIOD           8
#define SYNFEI_REQUIRED_PACKETS 20
#define SYNFEI_MAX_DIFFERENCE   4

void updateSynFei()
{
    if (synFei_PacketReceivedNonce == -1)
        return;

    uint8_t receivedNonce = synFei_PacketReceivedNonce % SYNFEI_PERIOD;
    synFei_PacketReceivedNonce = -1;

    static uint8_t synFei_cntHi;
    static uint8_t synFei_cntLo;

    if (receivedNonce == 1)
    {
        ++synFei_cntHi;
    }
    else if (receivedNonce == 5)
    {
        ++synFei_cntLo;
    }

    if (synFei_cntHi + synFei_cntLo >= SYNFEI_REQUIRED_PACKETS)
    {
        int32_t diff = (int32_t)synFei_cntHi - (int32_t)synFei_cntLo;
        DBGLN("synFei=%d", diff);
        if (diff >= SYNFEI_MAX_DIFFERENCE)
            HandleFreqCorr(false);
        else if (diff <= -SYNFEI_MAX_DIFFERENCE)
            HandleFreqCorr(true);

        synFei_cntHi = synFei_cntLo = 0;
    }
}
```

Even at a regular LQ of ~60%, these values seem to be really good at rejecting false positives. An hour of 500Hz 60LQ with a receiver that has an offset of 31 normally had ZERO false positives. The minimum adjustment rate for one step of Frequency Correction is:
| Rate | Min Correction Rate(Hz) | 
|--|--|
| F1000 / Dmodes | 6.25 |
| F500 / 500 | 3.125 |
| 250 | 1.5625 |
| 150 | 0.9375 |
| 50 | 0.3125 |

These rates are sufficient for all packet rates to be able to counter temperature-induced frequency offset. I think that's what we're trying to avoid? Correct me if I am wrong about the purpose of all of this.

## Disadvantages
* Only affects the worst of the worst of the receivers. Your receiver has to be pretty much on the edge of not working for this to do anything. The frequency shift a pilot might experience due to temperature is less than the shift used to perform this measurement, so without this code in place, that pilot would _probably_ not have experienced any decreased link integrity. It just can be overstated how close to the edge of tolerance the system needs to be for this to have any positive effect.
* It never "centers" on the center frequency of the TX, like the hardware FEI does. It just gets slightly better to the point that the packet loss is reduced under the threshold.
* Link Quality is actually decreased for those who were at 100LQ before, but their receivers were close to the tolerance limit. They would have been fine but now 1/8th of their packets are getting lost and it takes time for the frequency to correct itself. This will create a lot of confusion why the LQ starts low and then after 20-30 seconds is ok.
* A user may not ever know their RX is out of tolerance. This just reduces the failsafe chance due to temperature drift so they'll happily just keep using shit hardware riding on the very edge of not working. There is no way to communicate to the user that their RX is out of whack and should not be used. Even if we get the flag back to the TX side, absolutely no one will go into Lua **after** they've flown for a while and before they plug in another model. Even if they do, they will also completely ignore the flashing message or popup saying "Receiver BAD CRYSTAL". Users will continue to use their bad hardware with suboptimal performance.
* This now makes our channels overlap slightly. This will decrease coexistence LQ in adjacent channels when flying with other pilots.
* If the user flies until they failsafe due to range, our current code resets the FreqCorrection and it is likely they won't be able to reestablish a connection. The frequency shifts in flight, the system adapted to it, and the RX is now outside tolerance and can't see packets from the TX any more. The correction resets on disconnect as a safety measure, in case the correction was what caused the failsafe, so eliminating this entirely may not be the safest solution for everyone. We can consider maybe just reducing the correction on disconnect though, so this is probably 

## Conclusions
This is a really clever procedure, but I'm really not sure there's much benefit to it. It adds complexity benefiting a very small number of users, and actually reduces performance for some users whose hardware seemed ok before, and reduces LQ when flying with others. Those who would benefit from this will likely never know they shouldn't be using their hardware.

I do not have the code saved to any branch, as it currently requires running with no telemetry due to the difficulty handling that will be. The TX needs to shift its frequency after the telemetry is received but long enough before the subsequent transmit that the frequency change can take effect. If the telemetry packet is missed, there is no event available to perform the frequency change in advance of the TOCK that will be used to initiate transmit.

EDIT: Not that this doesn't have merit as a last-ditch-effort to prevent a failsafe on the worst hardware though, so maybe some version of this can be integrated. The largest concern being each TX now occupying two channels at once.

**Reply by SunjunKim** [reply]
> * Only affects the worst of the worst of the receivers. Your receiver has to be pretty much on the edge of not working for this to do anything. The frequency shift a pilot might experience due to temperature is less than the shift used to perform this measurement, so without this code in place, that pilot would _probably_ not have experienced any decreased link integrity. It just can be overstated how close to the edge of tolerance the system needs to be for this to have any positive effect.

This approach is indeed for the TX-RX pair on the edge – which barely able to bind but might be disconnected during the flight due to the temperature-induced frequency shift.

> * It never "centers" on the center frequency of the TX, like the hardware FEI does. It just gets slightly better to the point that the packet loss is reduced under the threshold.

The purpose of this is not to make the link fully disconnected. So if this makes the link alive (even with low LQ) with the frequency offset out of tolerance, it's done its job.

> * Link Quality is actually decreased for those who were at 100LQ before, but their receivers were close to the tolerance limit. They would have been fine but now 1/8th of their packets are getting lost and it takes time for the frequency to correct itself. This will create a lot of confusion why the LQ starts low and then after 20-30 seconds is ok.
> * A user may not ever know their RX is out of tolerance. This just reduces the failsafe chance due to temperature drift so they'll happily just keep using shit hardware riding on the very edge of not working. There is no way to communicate to the user that their RX is out of whack and should not be used. Even if we get the flag back to the TX side, absolutely no one will go into Lua **after** they've flown for a while and before they plug in another model. Even if they do, they will also completely ignore the flashing message or popup saying "Receiver BAD CRYSTAL". Users will continue to use their bad hardware with suboptimal performance.

We should really think about a better way to notify a user and be transparent about the problem in the hardware, which will also reduce the confusion you mentioned. I don't have a silver bullet for this problem for now though.

> * This now makes our channels overlap slightly. This will decrease coexistence LQ in adjacent channels when flying with other pilots.

An RX-only implementation could solve this.

> * If the user flies until they failsafe due to range, our current code resets the FreqCorrection and it is likely they won't be able to reestablish a connection. The frequency shifts in flight, the system adapted to it, and the RX is now outside tolerance and can't see packets from the TX any more. The correction resets on disconnect as a safety measure, in case the correction was what caused the failsafe, so eliminating this entirely may not be the safest solution for everyone. We can consider maybe just reducing the correction on disconnect though, so this is probably

It's related to the "how to notify users that they have shitty hardware" problem. I'll think about the solution and propose it If I come up with any good idea.

**Reply by SunjunKim** [reply]
> If the user flies until they failsafe due to range, our current code resets the FreqCorrection and it is likely they won't be able to reestablish a connection. The frequency shifts in flight, the system adapted to it, and the RX is now outside tolerance and can't see packets from the TX any more. The correction resets on disconnect as a safety measure, in case the correction was what caused the failsafe, so eliminating this entirely may not be the safest solution for everyone. We can consider maybe just reducing the correction on disconnect though, so this is probably

Some random idea

1. Case: correction is reset every time. 

Whenever an RX experiences the frequency compensation, flag some bit in EEPROM, and power it up with a "warning pattern" LED in the next booting? If it's connected it'll do a correction and show a solid LED. If it doesn't, it'll show some warning message (=LED) to a user.

2. Case: RX saves the correction

Whenever an RX experiences the frequency compensation, save the amount of frequency compensated. In the next booting, cycle through RF modes once without and once with the corrected frequency.

**Comment by robustini** [comment]
Yes, the heat problem is terrible.
If, as it is usually done, you put a heat shrink tubing on the rx it is dramatic.
With the hair dryer I only speeded up the thing, but with the heat shrink tubing it does it after about a minute starting at an ambient temperature of 25°.

https://youtu.be/Cp5V1GTqNBI

**Comment by CapnBry** [comment]
> With the hair dryer I only speeded up the thing, but with the heat shrink tubing it does it after about a minute starting at an ambient temperature of 25°.

That's an R9MM which already has fully working and fast-correcting frequency correction. It disconnected because it overheated beyond correctable tolerance, or a component got too hot to function properly. That's not relevant to this discussion which is trying to **add** frequency correction.

**Reply by robustini** [reply]
I might agree with you if I haven't seen it working with ACCESS (F-port with telemetry) in an hour and closed in shrink wrap.
So is it possible that FrSky has an even more incisive frequency correction?
The temperature on the rx is about the same.
